rstudio/shinyloadtest
# session$registerDataObj() support (formerly: DT support)

This is the R side of https://github.com/rstudio/shinycannon/pull/46.

## Dev tasks

- [x] Rename `DT_NONCE_<n>` to `SHINY_DATAOBJ_NONCE_<n>` or similar to indicate the nonce business is from Shiny, not from DT
- [x] Test dynamic UI
- [x] Test Selectize with `server = TRUE`

## Background

Previously, we *thought* we supported DT &mdash; and other inputs/outputs that used XHR POST requests, like Selectize &mdash; ever since we acquired POST support. But that turned out not to be true, because the URLs being POSTed to contained random session data that would cause shinycannon playback to fail.

Here's an example from a recording of the kind of POST we'd record, but that couldn't be played back:
```
{
   "type":"REQ_POST",
   "begin":"2019-11-14T23:56:08.112Z",
   "end":"2019-11-14T23:56:08.133Z",
   "status":200,
   "url":"/session/${SESSION}/dataobj/rawdata?w=&nonce=460858f4e5fbb54e",
   "datafile":"021-selectize-plot.log.post.0"
}
```
Notice that `nonce=460858f4e5fbb54e` part of the `"url"` field. The `nonce` value is something that was sent to the client earlier as part of a websocket message body sent from the server to the browser. Because it's baked into the recording, and not understood as a session-specific parameter like `${SESSION}`, replaying the post during playback will not work.

## Strategy

The original idea of this PR was to specifically recognize and parse DT websocket messages that contained the `nonce` value, and to parse out the `nonce` value and recognize it as a session-specific parameter like `${SESSION}`.

However, extending support specifically to DT turned out to be a waste of effort, considering the `nonce` machinery is only incidentally related to DT. Really, any R code that used `session$registerDataObj()` was susceptible to this problem, and so the best solution would be more general.

So, instead of dealing specifically with DT, I generalized the code to a general regex scan of incoming messages for URLs with `nonce`s that look like the kind generated by `session$registerDataObj()`.

When they are found, they are replaced with `${SHINY_DATAOBJ_NONCE_<N>}` placeholder values. The big difference between these type of placeholders and any other kind we already have is that there can be many of them per session, since an application or its packages might call `session$registerDataObj()` an arbitrary number of times.

## QA Notes

After this (and with https://github.com/rstudio/shinycannon/pull/46) it should be possible to record and load test the following example applications without error:

* 021-selectize-plot
* 063-superzip-example - Using the "Data Explorer" tab during recording shouldn't break recording or cause playback problems
* 140-selectize-inputs - All of the inputs should record properly. Server-side inputs should not pose a problem.
